File: app/macos/Runner/NativePlaybackTexture.swift
import Cocoa
import CoreVideo
import Darwin
import FlutterMacOS

final class NativePlaybackTexture: NSObject, FlutterTexture {
  private weak var registry: FlutterTextureRegistry?
  private(set) var textureId: Int64 = -1

  private var libHandle: UnsafeMutableRawPointer?
  private var session: UnsafeMutableRawPointer?

  private var pixelBuffer: CVPixelBuffer?
  private var bufferWidth = 0
  private var bufferHeight = 0
  private var displayTimer: Timer?

  typealias CreateFn = @convention(c) (UnsafePointer<CChar>?) -> UnsafeMutableRawPointer?
  typealias DestroyFn = @convention(c) (UnsafeMutableRawPointer?) -> Void
  typealias StartFn = @convention(c) (UnsafeMutableRawPointer?) -> Int32
  typealias PauseFn = @convention(c) (UnsafeMutableRawPointer?) -> Void
  typealias IsPlayingFn = @convention(c) (UnsafeMutableRawPointer?) -> Int32
  typealias SeekFn = @convention(c) (UnsafeMutableRawPointer?, Double) -> Int32
  typealias PositionFn = @convention(c) (UnsafeMutableRawPointer?) -> Double
  typealias DurationFn = @convention(c) (UnsafeMutableRawPointer?) -> Double
  typealias FpsFn = @convention(c) (UnsafeMutableRawPointer?) -> Double
  typealias LockFrameFn = @convention(c) (
    UnsafeMutableRawPointer?,
    UnsafeMutablePointer<Int32>?,
    UnsafeMutablePointer<Int32>?
  ) -> UnsafePointer<UInt8>?
  typealias UnlockFrameFn = @convention(c) (UnsafeMutableRawPointer?) -> Void
  typealias ReachedEndFn = @convention(c) (UnsafeMutableRawPointer?) -> Int32

  private var fCreate: CreateFn?
  private var fDestroy: DestroyFn?
  private var fStart: StartFn?
  private var fPause: PauseFn?
  private var fResume: PauseFn?
  private var fIsPlaying: IsPlayingFn?
  private var fSeek: SeekFn?
  private var fPosition: PositionFn?
  private var fDuration: DurationFn?
  private var fFps: FpsFn?
  private var fLockFrame: LockFrameFn?
  private var fUnlockFrame: UnlockFrameFn?
  private var fReachedEnd: ReachedEndFn?

  init(registry: FlutterTextureRegistry) {
    self.registry = registry
    super.init()
  }

  // MARK: - Lifecycle

  func open(engineLibPath: String, mediaPath: String, result: @escaping FlutterResult) {
    close()

    guard let handle = dlopen(engineLibPath, RTLD_NOW | RTLD_LOCAL) else {
      result(FlutterError(
        code: "dlopen_failed",
        message: String(cString: dlerror()),
        details: nil))
      return
    }
    libHandle = handle
    loadSymbols(from: handle)

    guard fCreate != nil, fDestroy != nil, fStart != nil else {
      result(FlutterError(
        code: "abi_error",
        message: "engine missing playback symbols — rebuild engine",
        details: nil))
      return
    }

    guard let sessionPtr = fCreate?(mediaPath) else {
      result(FlutterError(
        code: "open_failed",
        message: "could not open media for playback",
        details: nil))
      return
    }
    session = sessionPtr

    textureId = registry?.register(self) ?? -1
    if textureId < 0 {
      result(FlutterError(
        code: "register_failed",
        message: "texture registration failed",
        details: nil))
      return
    }

    _ = fStart?(session)
    startDisplayTimer()

    result([
      "textureId": textureId,
      "duration": fDuration?(session) ?? 0,
      "fps": displayFrameRate,
    ])
  }

  func close() {
    displayTimer?.invalidate()
    displayTimer = nil

    if let session = session {
      fPause?(session)
      fDestroy?(session)
    }
    session = nil

    if textureId >= 0, let registry = registry {
      registry.unregisterTexture(textureId)
    }
    textureId = -1
    pixelBuffer = nil
    bufferWidth = 0
    bufferHeight = 0

    if let handle = libHandle {
      dlclose(handle)
      libHandle = nil
    }
  }

  deinit {
    close()
  }

  // MARK: - Playback controls

  func play() {
    guard let session = session else { return }
    if fIsPlaying?(session) == 0 {
      if reachedEnd() {
        _ = fSeek?(session, 0.0)
      }
      fResume?(session)
    } else {
      _ = fStart?(session)
    }
  }

  func pause() {
    guard let session = session else { return }
    fPause?(session)
  }

  func seek(_ seconds: Double) {
    guard let session = session else { return }
    _ = fSeek?(session, seconds)
  }

  func position() -> Double {
    guard let session = session else { return 0 }
    return fPosition?(session) ?? 0
  }

  func duration() -> Double {
    guard let session = session else { return 0 }
    return fDuration?(session) ?? 0
  }

  func reachedEnd() -> Bool {
    guard let session = session, let f = fReachedEnd else { return false }
    return f(session) == 1
  }

  func isPlaying() -> Bool {
    guard let session = session, let f = fIsPlaying else { return false }
    return f(session) == 1
  }

  // MARK: - Engine symbols

  private func loadSymbols(from handle: UnsafeMutableRawPointer) {
    func sym<T>(_ name: String, as type: T.Type) -> T? {
      guard let p = dlsym(handle, name) else { return nil }
      return unsafeBitCast(p, to: type)
    }

    fCreate = sym("cc_playback_create", as: CreateFn.self)
    fDestroy = sym("cc_playback_destroy", as: DestroyFn.self)
    fStart = sym("cc_playback_start", as: StartFn.self)
    fPause = sym("cc_playback_pause", as: PauseFn.self)
    fResume = sym("cc_playback_resume", as: PauseFn.self)
    fIsPlaying = sym("cc_playback_is_playing", as: IsPlayingFn.self)
    fSeek = sym("cc_playback_seek", as: SeekFn.self)
    fPosition = sym("cc_playback_position", as: PositionFn.self)
    fDuration = sym("cc_playback_duration", as: DurationFn.self)
    fFps = sym("cc_playback_fps", as: FpsFn.self)
    fLockFrame = sym("cc_playback_lock_frame", as: LockFrameFn.self)
    fUnlockFrame = sym("cc_playback_unlock_frame", as: UnlockFrameFn.self)
    fReachedEnd = sym("cc_playback_reached_end", as: ReachedEndFn.self)
  }

  // MARK: - Frame delivery

  private var displayFrameRate: Double {
    min(max(fFps?(session) ?? 30.0, 10), 60)
  }

  private func startDisplayTimer() {
    let timer = Timer(timeInterval: 1.0 / displayFrameRate, repeats: true) { [weak self] _ in
      self?.presentTick()
    }
    RunLoop.main.add(timer, forMode: .common)
    displayTimer = timer
  }

  private func presentTick() {
    guard let session = session,
          let lock = fLockFrame,
          let unlock = fUnlockFrame,
          let registry = registry else { return }

    var width: Int32 = 0
    var height: Int32 = 0
    guard let pixels = lock(session, &width, &height), width > 0, height > 0 else {
      return
    }
    defer { unlock(session) }

    ensurePixelBuffer(width: width, height: height)
    guard let pixelBuffer = pixelBuffer else { return }

    CVPixelBufferLockBaseAddress(pixelBuffer, [])
    defer { CVPixelBufferUnlockBaseAddress(pixelBuffer, []) }
    copyFrame(
      pixels: pixels,
      width: Int(width),
      height: Int(height),
      into: pixelBuffer)
    registry.textureFrameAvailable(textureId)
  }

  private func ensurePixelBuffer(width: Int32, height: Int32) {
    guard pixelBuffer == nil || Int(width) != bufferWidth || Int(height) != bufferHeight else {
      return
    }

    var newBuffer: CVPixelBuffer?
    let attrs: [CFString: Any] = [
      kCVPixelBufferPixelFormatTypeKey: kCVPixelFormatType_32RGBA,
      kCVPixelBufferWidthKey: Int(width),
      kCVPixelBufferHeightKey: Int(height),
    ]
    guard CVPixelBufferCreate(
      kCFAllocatorDefault,
      Int(width),
      Int(height),
      kCVPixelFormatType_32RGBA,
      attrs as CFDictionary,
      &newBuffer) == kCVReturnSuccess,
      let created = newBuffer else {
      return
    }

    pixelBuffer = created
    bufferWidth = Int(width)
    bufferHeight = Int(height)
  }

  private func copyFrame(
    pixels: UnsafePointer<UInt8>,
    width: Int,
    height: Int,
    into pixelBuffer: CVPixelBuffer
  ) {
    guard let base = CVPixelBufferGetBaseAddress(pixelBuffer) else { return }

    let destinationStride = CVPixelBufferGetBytesPerRow(pixelBuffer)
    let sourceStride = width * 4
    let destination = base.assumingMemoryBound(to: UInt8.self)
    for row in 0..<height {
      memcpy(
        destination + row * destinationStride,
        pixels + row * sourceStride,
        min(sourceStride, destinationStride))
    }
  }

  func copyPixelBuffer() -> Unmanaged<CVPixelBuffer>? {
    guard let pixelBuffer = pixelBuffer else { return nil }
    return Unmanaged.passRetained(pixelBuffer)
  }
}

File: app/windows/runner/native_playback_texture.cpp
#include "native_playback_texture.h"

#include <algorithm>

namespace {

template <typename T>
T LoadSym(HMODULE module, const char* name) {
  return reinterpret_cast<T>(GetProcAddress(module, name));
}

}  // namespace

NativePlaybackTexture::NativePlaybackTexture(
    flutter::TextureRegistrar* texture_registrar)
    : texture_registrar_(texture_registrar) {}

NativePlaybackTexture::~NativePlaybackTexture() { Close(); }

bool NativePlaybackTexture::LoadSymbols() {
  f_create_ = LoadSym<CreateFn>(lib_handle_, "cc_playback_create");
  f_destroy_ = LoadSym<DestroyFn>(lib_handle_, "cc_playback_destroy");
  f_start_ = LoadSym<StartFn>(lib_handle_, "cc_playback_start");
  f_pause_ = LoadSym<PauseResumeFn>(lib_handle_, "cc_playback_pause");
  f_resume_ = LoadSym<PauseResumeFn>(lib_handle_, "cc_playback_resume");
  f_is_playing_ = LoadSym<IsPlayingFn>(lib_handle_, "cc_playback_is_playing");
  f_seek_ = LoadSym<SeekFn>(lib_handle_, "cc_playback_seek");
  f_position_ = LoadSym<PositionFn>(lib_handle_, "cc_playback_position");
  f_duration_ = LoadSym<DurationFn>(lib_handle_, "cc_playback_duration");
  f_fps_ = LoadSym<FpsFn>(lib_handle_, "cc_playback_fps");
  f_lock_frame_ = LoadSym<LockFrameFn>(lib_handle_, "cc_playback_lock_frame");
  f_unlock_frame_ =
      LoadSym<UnlockFrameFn>(lib_handle_, "cc_playback_unlock_frame");
  f_reached_end_ =
      LoadSym<ReachedEndFn>(lib_handle_, "cc_playback_reached_end");
  return f_create_ && f_destroy_ && f_start_;
}

void NativePlaybackTexture::Open(
    const std::string& engine_lib_path, const std::string& media_path,
    std::unique_ptr<flutter::MethodResult<flutter::EncodableValue>> result) {
  Close();

  lib_handle_ = LoadLibraryA(engine_lib_path.c_str());
  if (!lib_handle_) {
    result->Error("dlopen_failed",
                   "LoadLibrary failed with error " +
                       std::to_string(GetLastError()));
    return;
  }

  if (!LoadSymbols()) {
    result->Error("abi_error",
                   "engine missing playback symbols \xe2\x80\x94 rebuild engine");
    return;
  }

  session_ = f_create_(media_path.c_str());
  if (!session_) {
    result->Error("open_failed", "could not open media for playback");
    return;
  }

  texture_ = std::make_unique<flutter::TextureVariant>(flutter::PixelBufferTexture(
      [this](size_t width, size_t height) -> const FlutterDesktopPixelBuffer* {
        return CopyPixelBuffer(width, height);
      }));
  texture_id_ = texture_registrar_->RegisterTexture(texture_.get());
  if (texture_id_ < 0) {
    result->Error("register_failed", "texture registration failed");
    return;
  }

  f_start_(session_);

  running_ = true;
  present_thread_ = std::thread([this] { PresentLoop(); });

  flutter::EncodableMap response{
      {flutter::EncodableValue("textureId"),
       flutter::EncodableValue(texture_id_)},
      {flutter::EncodableValue("duration"),
       flutter::EncodableValue(f_duration_ ? f_duration_(session_) : 0.0)},
      {flutter::EncodableValue("fps"), flutter::EncodableValue(DisplayFrameRate())},
  };
  result->Success(flutter::EncodableValue(response));
}

double NativePlaybackTexture::DisplayFrameRate() {
  double fps = f_fps_ ? f_fps_(session_) : 30.0;
  return std::min(std::max(fps, 10.0), 60.0);
}

void NativePlaybackTexture::PresentLoop() {
  const auto interval =
      std::chrono::duration<double>(1.0 / DisplayFrameRate());
  while (running_.load()) {
    if (session_ && f_lock_frame_ && f_unlock_frame_) {
      int32_t width = 0, height = 0;
      const uint8_t* pixels = f_lock_frame_(session_, &width, &height);
      if (pixels && width > 0 && height > 0) {
        int back = 1 - front_index_.load();
        auto& buf = buffers_[back];
        const size_t needed = static_cast<size_t>(width) * height * 4;
        if (buf.data.size() != needed) {
          std::lock_guard<std::mutex> lock(resize_mutex_);
          buf.data.resize(needed);
        }
        buf.width = static_cast<size_t>(width);
        buf.height = static_cast<size_t>(height);
        std::copy(pixels, pixels + needed, buf.data.begin());
        front_index_.store(back);
        texture_registrar_->MarkTextureFrameAvailable(texture_id_);
      }
      f_unlock_frame_(session_);
    }
    std::this_thread::sleep_for(
        std::chrono::duration_cast<std::chrono::milliseconds>(interval));
  }
}

const FlutterDesktopPixelBuffer* NativePlaybackTexture::CopyPixelBuffer(
    size_t /*width*/, size_t /*height*/) {
  auto& buf = buffers_[front_index_.load()];
  if (buf.data.empty()) return nullptr;
  pixel_buffer_.buffer = buf.data.data();
  pixel_buffer_.width = buf.width;
  pixel_buffer_.height = buf.height;
  return &pixel_buffer_;
}

bool NativePlaybackTexture::ReachedEnd() {
  if (!session_ || !f_reached_end_) return false;
  return f_reached_end_(session_) == 1;
}

void NativePlaybackTexture::Play() {
  if (!session_) return;
  if (f_is_playing_ && f_is_playing_(session_) == 0) {
    if (ReachedEnd() && f_seek_) f_seek_(session_, 0.0);
    if (f_resume_) f_resume_(session_);
  } else if (f_start_) {
    f_start_(session_);
  }
}

void NativePlaybackTexture::Pause() {
  if (session_ && f_pause_) f_pause_(session_);
}

void NativePlaybackTexture::Seek(double seconds) {
  if (session_ && f_seek_) f_seek_(session_, seconds);
}

double NativePlaybackTexture::Position() {
  if (!session_ || !f_position_) return 0;
  return f_position_(session_);
}

bool NativePlaybackTexture::IsPlaying() {
  if (!session_ || !f_is_playing_) return false;
  return f_is_playing_(session_) == 1;
}

void NativePlaybackTexture::Close() {
  running_ = false;
  if (present_thread_.joinable()) present_thread_.join();

  if (session_) {
    if (f_pause_) f_pause_(session_);
    if (f_destroy_) f_destroy_(session_);
  }
  session_ = nullptr;

  if (texture_id_ >= 0) {
    texture_registrar_->UnregisterTexture(texture_id_);
    texture_id_ = -1;
  }
  texture_.reset();

  buffers_[0] = FrameBuffer{};
  buffers_[1] = FrameBuffer{};
  front_index_ = 0;

  if (lib_handle_) {
    FreeLibrary(lib_handle_);
    lib_handle_ = nullptr;
  }
}
